File: scripts/hygiene/oracle-fence.mjs
#!/usr/bin/env bun
// oracle-fence — T5 (docs/testing/rules.md): "Oracles never import the
// kernel. The from-scratch evaluator and the fixture harness live in a
// separate build target whose include path cannot reach the extension's
// sources; a meta-test walks includes and fails on any edge." AGENTS.md's
// own "Where things are" table states the same rule for this repo:
// "test/oracle/ — the from-scratch oracle (must never include `src/`)."
//
// Walks every `#include "..."` (quoted, relative — never `#include <...>`,
// which is always a standard/system header and never a path in this
// repository) reachable from every file under `test/oracle/`, transitively,
// resolving each include relative to the including file's own directory
// (matching how a C++ compiler itself resolves a quoted include), and fails
// if any resolved path lands under `src/`. This is the actual "walk includes
// and fails on any edge" meta-test T5 asks for — a real graph walk, not a
// single-directory grep, so a `test/oracle/a.hpp` that includes a
// `test/oracle/b.hpp` that includes `src/kernel/comparator.hpp` is still
// caught even though `a.hpp` itself never spells `src/` anywhere.
import { readFileSync, readdirSync, statSync, existsSync } from "node:fs";
import { join, dirname, normalize, relative, sep } from "node:path";

const args = process.argv.slice(2);
const rootIdx = args.indexOf("--root");
const root = rootIdx === -1 ? process.cwd() : args[rootIdx + 1];

const ORACLE_DIR = join(root, "test", "oracle");
const INCLUDE_RE = /^\s*#\s*include\s*"([^"]+)"/gm;
const CPP_EXT_RE = /\.(hpp|hh|h|hxx|cpp|cc|cxx)$/;

function listFiles(dir) {
  if (!existsSync(dir)) return [];
  const out = [];
  const walk = (d) => {
    for (const entry of readdirSync(d).sort()) {
      const full = join(d, entry);
      const st = statSync(full);
      if (st.isDirectory()) walk(full);
      else if (CPP_EXT_RE.test(entry)) out.push(full);
    }
  };
  walk(dir);
  return out;
}

function quotedIncludes(path) {
  const content = readFileSync(path, "utf8");
  const out = [];
  let m;
  INCLUDE_RE.lastIndex = 0;
  while ((m = INCLUDE_RE.exec(content))) out.push(m[1]);
  return out;
}

function toPosixRelative(fromRoot) {
  return relative(root, fromRoot).split(sep).join("/");
}

const oracleFiles = listFiles(ORACLE_DIR);
if (oracleFiles.length === 0) {
  console.error("oracle-fence: FAIL");
  console.error(`  test/oracle has no .hpp/.cpp files to fence (Article V.2 / AGENTS.md: the from-scratch oracle lives there)`);
  process.exit(1);
}

const violations = [];
const visited = new Set();

function walk(absPath, chain) {
  const key = normalize(absPath);
  if (visited.has(key)) return;
  visited.add(key);
  if (!existsSync(absPath)) return; // an include this scan can't resolve — nothing further to walk

  for (const inc of quotedIncludes(absPath)) {
    const resolved = normalize(join(dirname(absPath), inc));
    const relPath = toPosixRelative(resolved);
    const nextChain = [...chain, relPath];
    if (relPath === "src" || relPath.startsWith("src/")) {
      violations.push(`${chain[0]}: reaches ${relPath} via ${nextChain.join(" -> ")}`);
      continue; // don't also walk into src/ — the violation is already recorded
    }
    walk(resolved, nextChain);
  }
}

for (const file of oracleFiles) {
  const rel = toPosixRelative(file);
  walk(file, [rel]);
}

if (violations.length > 0) {
  console.error("oracle-fence: FAIL");
  for (const v of violations) console.error(`  ${v}`);
  process.exit(1);
}
console.log(`oracle-fence: PASS (${oracleFiles.length} file(s) under test/oracle, no edge reaches src/)`);

File: scripts/hygiene/oracle-fence.py
#!/usr/bin/env python3
# oracle-fence — T5 (docs/testing/rules.md): "Oracles never import the
# kernel. The from-scratch evaluator and the fixture harness live in a
# separate build target whose include path cannot reach the extension's
# sources; a meta-test walks includes and fails on any edge." AGENTS.md's
# own "Where things are" table states the same rule for this repo:
# "test/oracle/ — the from-scratch oracle (must never include `src/`)."
#
# Walks every `#include "..."` (quoted, relative — never `#include <...>`,
# which is always a standard/system header and never a path in this
# repository) reachable from every file under `test/oracle/`, transitively,
# resolving each include relative to the including file's own directory
# (matching how a C++ compiler itself resolves a quoted include), and fails
# if any resolved path lands under `src/`.
import os
import re
import sys

args = sys.argv[1:]
root = args[args.index("--root") + 1] if "--root" in args else os.getcwd()

ORACLE_DIR = os.path.join(root, "test", "oracle")
INCLUDE_RE = re.compile(r'^\s*#\s*include\s*"([^"]+)"', re.MULTILINE)
CPP_EXT_RE = re.compile(r"\.(hpp|hh|h|hxx|cpp|cc|cxx)$")


def list_files(dir_):
    if not os.path.isdir(dir_):
        return []
    out = []

    def walk(d):
        for entry in sorted(os.listdir(d)):
            full = os.path.join(d, entry)
            if os.path.isdir(full):
                walk(full)
            elif CPP_EXT_RE.search(entry):
                out.append(full)

    walk(dir_)
    return out


def quoted_includes(path):
    with open(path, "r", encoding="utf-8") as f:
        content = f.read()
    return INCLUDE_RE.findall(content)


def to_posix_relative(from_root):
    return os.path.relpath(from_root, root).replace(os.sep, "/")


oracle_files = list_files(ORACLE_DIR)
if not oracle_files:
    print("oracle-fence: FAIL", file=sys.stderr)
    print(
        "  test/oracle has no .hpp/.cpp files to fence (Article V.2 / AGENTS.md: the from-scratch oracle lives there)",
        file=sys.stderr,
    )
    sys.exit(1)

violations = []
visited = set()


def walk(abs_path, chain):
    key = os.path.normpath(abs_path)
    if key in visited:
        return
    visited.add(key)
    if not os.path.exists(abs_path):
        return

    for inc in quoted_includes(abs_path):
        resolved = os.path.normpath(os.path.join(os.path.dirname(abs_path), inc))
        rel_path = to_posix_relative(resolved)
        next_chain = chain + [rel_path]
        if rel_path == "src" or rel_path.startswith("src/"):
            violations.append(f"{chain[0]}: reaches {rel_path} via {' -> '.join(next_chain)}")
            continue
        walk(resolved, next_chain)


for file in oracle_files:
    rel = to_posix_relative(file)
    walk(file, [rel])

if violations:
    print("oracle-fence: FAIL", file=sys.stderr)
    for v in violations:
        print(f"  {v}", file=sys.stderr)
    sys.exit(1)
print(f"oracle-fence: PASS ({len(oracle_files)} file(s) under test/oracle, no edge reaches src/)")
